# Experiment 6: Max-Min Composition and Max-Product Composition
**Name:** Arshan Attar  
**Roll Number:** 231408  

In [ ]:
import numpy as np

# Max-Min Composition given by Zadeh
def maxMin(x, y):
    z = []
    for x1 in x:
        for y1 in y.T:
            z.append(max(np.minimum(x1, y1)))
    return np.array(z).reshape((x.shape[0], y.shape[1]))

# Max-Product Composition given by Rosenfeld
def maxProduct(x, y):
    z = []
    for x1 in x:
        for y1 in y.T:
            z.append(max(np.multiply(x1, y1)))
    return np.array(z).reshape((x.shape[0], y.shape[1]))

# 3 arrays for the example
r1 = np.array([
    [0.9, 0.2, 0.5],
    [0.4, 0.8, 0.1],
    [0.3, 0.6, 1.0]
])

print("R1:\n", r1)

r2 = np.array([
    [0.7, 0.3, 0.6],
    [0.2, 0.9, 0.4],
    [0.5, 0.1, 0.8]
])

print("R2:\n", r2)

r3 = np.array([
    [1.0, 0.2, 0.4],
    [0.3, 1.0, 0.5],
    [0.6, 0.4, 1.0]
])

print("R3:\n", r3)

print("\n")

print("R1oR2 => Max-Min :\n" + str(maxMin(r1, r2)) + "\n")
print("R1oR2 => Max-Product :\n" + str(maxProduct(r1, r2)) + "\n\n")

print("R1oR3 => Max-Min :\n" + str(maxMin(r1, r3)) + "\n")
print("R1oR3 => Max-Product :\n" + str(maxProduct(r1, r3)) + "\n\n")

print("R1oR2oR3 => Max-Min :\n" + str(maxMin(r1, maxMin(r2, r3))) + "\n")
print("R1oR2oR3 => Max-Product :\n" + str(maxProduct(r1, maxProduct(r2, r3))) + "\n\n")

R1:
 [[0.9 0.2 0.5]
 [0.4 0.8 0.1]
 [0.3 0.6 1. ]]
R2:
 [[0.7 0.3 0.6]
 [0.2 0.9 0.4]
 [0.5 0.1 0.8]]
R3:
 [[1.  0.2 0.4]
 [0.3 1.  0.5]
 [0.6 0.4 1. ]]


R1oR2 => Max-Min :
[[0.7 0.3 0.6]
 [0.4 0.8 0.4]
 [0.5 0.6 0.8]]

R1oR2 => Max-Product :
[[0.63 0.27 0.54]
 [0.28 0.72 0.32]
 [0.5  0.54 0.8 ]]


R1oR3 => Max-Min :
[[0.9 0.4 0.5]
 [0.4 0.8 0.5]
 [0.6 0.6 1. ]]

R1oR3 => Max-Product :
[[0.9 0.2 0.5]
 [0.4 0.8 0.4]
 [0.6 0.6 1. ]]


R1oR2oR3 => Max-Min :
[[0.7 0.4 0.6]
 [0.4 0.8 0.5]
 [0.6 0.6 0.8]]

R1oR2oR3 => Max-Product :
[[0.63 0.27 0.54]
 [0.28 0.72 0.36]
 [0.5  0.54 0.8 ]]




## Code Explanation

This code utilizes the `numpy` library for efficient array operations to demonstrate two ways to combine fuzzy matrices (relations) using composition:

1.  **Max-Min Composition (`Zadeh's Method`)**:
    This is the standard method in fuzzy logic. It evaluates the "weakest link" (minimum) in each possible chain of relationships, and then chooses the "strongest" (maximum) of those weakest links.

2.  **Max-Product Composition (`Rosenfeld's Method`)**:
    Instead of just looking at the minimum, this method multiplies the probabilities (or degrees of membership). This makes it more sensitive to small values, as a small number multiplied by another small number becomes even smaller.

### How the Functions Work
The custom functions, `maxMin` and `maxProduct`, take two matrices (`x` and `y`) as inputs:

*   **The Loops**: Nested `for` loops iterate through each row of the first matrix (`x`) and each column of the second matrix (accessed by transposing `y.T`).

*   **The Math**:
    *   `np.minimum(x1, y1)`: For Max-Min composition, this compares elements to save the smaller numbers.
    *   `np.multiply(x1, y1)`: For Max-Product composition, this multiplies the corresponding elements.

*   **Storing & Reshaping**: The `max(...)` function finds the highest number from the mathematical operation step and appends it to a list called `z`. Finally, `.reshape()` transforms that flat list back into a proper 2D matrix.

### Running the Experiment

1.  We create three 3x3 `numpy` arrays (`r1`, `r2`, and `r3`) with decimal values between 0.0 and 1.0, representing "fuzzy" weights.

2.  We print the initial matrices to observe the starting data.

3.  Finally, we print the results of combining two matrices at a time, as well as chaining all three together (e.g., `R1 o (R2 o R3)`). This vividly demonstrates how Zadeh's and Rosenfeld's methods yield different final answers, highlighting their distinct approaches to fuzzy composition.

# Fuzzy Matrix Composition: Manual Calculations

**Initial Matrices for the Experiment:**
$$R_1 = \begin{bmatrix} 0.9 & 0.2 & 0.5 \\ 0.4 & 0.8 & 0.1 \\ 0.3 & 0.6 & 1.0 \end{bmatrix}, \quad R_2 = \begin{bmatrix} 0.7 & 0.3 & 0.6 \\ 0.2 & 0.9 & 0.4 \\ 0.5 & 0.1 & 0.8 \end{bmatrix}, \quad R_3 = \begin{bmatrix} 1.0 & 0.2 & 0.4 \\ 0.3 & 1.0 & 0.5 \\ 0.6 & 0.4 & 1.0 \end{bmatrix}$$

---

## 1. Calculation: R1 o R2 (Max-Min Composition)
*Method: Pair row of R1 with column of R2, find the minimums, then select the maximum.*

**Row 1 of R1 against columns of R2:**
*   $C_{11} = \max(\min(0.9, 0.7), \min(0.2, 0.2), \min(0.5, 0.5)) = \max(0.7, 0.2, 0.5) = 0.7$
*   $C_{12} = \max(\min(0.9, 0.3), \min(0.2, 0.9), \min(0.5, 0.1)) = \max(0.3, 0.2, 0.1) = 0.3$
*   $C_{13} = \max(\min(0.9, 0.6), \min(0.2, 0.4), \min(0.5, 0.8)) = \max(0.6, 0.2, 0.5) = 0.6$

**Row 2 of R1 against columns of R2:**
*   $C_{21} = \max(\min(0.4, 0.7), \min(0.8, 0.2), \min(0.1, 0.5)) = \max(0.4, 0.2, 0.1) = 0.4$
*   $C_{22} = \max(\min(0.4, 0.3), \min(0.8, 0.9), \min(0.1, 0.1)) = \max(0.3, 0.8, 0.1) = 0.8$
*   $C_{23} = \max(\min(0.4, 0.6), \min(0.8, 0.4), \min(0.1, 0.8)) = \max(0.4, 0.4, 0.1) = 0.4$

**Row 3 of R1 against columns of R2:**
*   $C_{31} = \max(\min(0.3, 0.7), \min(0.6, 0.2), \min(1.0, 0.5)) = \max(0.3, 0.2, 0.5) = 0.5$
*   $C_{32} = \max(\min(0.3, 0.3), \min(0.6, 0.9), \min(1.0, 0.1)) = \max(0.3, 0.6, 0.1) = 0.6$
*   $C_{33} = \max(\min(0.3, 0.6), \min(0.6, 0.4), \min(1.0, 0.8)) = \max(0.3, 0.4, 0.8) = 0.8$

**Final Matrix for R1 o R2 (Max-Min):**
$$ \begin{bmatrix} 0.7 & 0.3 & 0.6 \\ 0.4 & 0.8 & 0.4 \\ 0.5 & 0.6 & 0.8 \end{bmatrix} $$

---

## 2. Calculation: R1 o R2 (Max-Product Composition)
*Method: Pair row of R1 with column of R2, multiply them, then select the maximum.*

**Row 1 of R1 against columns of R2:**
*   $P_{11} = \max((0.9 \times 0.7), (0.2 \times 0.2), (0.5 \times 0.5)) = \max(0.63, 0.04, 0.25) = 0.63$
*   $P_{12} = \max((0.9 \times 0.3), (0.2 \times 0.9), (0.5 \times 0.1)) = \max(0.27, 0.18, 0.05) = 0.27$
*   $P_{13} = \max((0.9 \times 0.6), (0.2 \times 0.4), (0.5 \times 0.8)) = \max(0.54, 0.08, 0.40) = 0.54$

**Row 2 of R1 against columns of R2:**
*   $P_{21} = \max((0.4 \times 0.7), (0.8 \times 0.2), (0.1 \times 0.5)) = \max(0.28, 0.16, 0.05) = 0.28$
*   $P_{22} = \max((0.4 \times 0.3), (0.8 \times 0.9), (0.1 \times 0.1)) = \max(0.12, 0.72, 0.01) = 0.72$
*   $P_{23} = \max((0.4 \times 0.6), (0.8 \times 0.4), (0.1 \times 0.8)) = \max(0.24, 0.32, 0.08) = 0.32$

**Row 3 of R1 against columns of R2:**
*   $P_{31} = \max((0.3 \times 0.7), (0.6 \times 0.2), (1.0 \times 0.5)) = \max(0.21, 0.12, 0.50) = 0.50$
*   $P_{32} = \max((0.3 \times 0.3), (0.6 \times 0.9), (1.0 \times 0.1)) = \max(0.09, 0.54, 0.10) = 0.54$
*   $P_{33} = \max((0.3 \times 0.6), (0.6 \times 0.4), (1.0 \times 0.8)) = \max(0.18, 0.24, 0.80) = 0.80$

**Final Matrix for R1 o R2 (Max-Product):**
$$ \begin{bmatrix} 0.63 & 0.27 & 0.54 \\ 0.28 & 0.72 & 0.32 \\ 0.50 & 0.54 & 0.80 \end{bmatrix} $$

---

## 3. Calculation: R1 o R3 (Max-Min Composition)

**Row 1 of R1 against columns of R3:**
*   $C_{11} = \max(\min(0.9, 1.0), \min(0.2, 0.3), \min(0.5, 0.6)) = \max(0.9, 0.2, 0.5) = 0.9$
*   $C_{12} = \max(\min(0.9, 0.2), \min(0.2, 1.0), \min(0.5, 0.4)) = \max(0.2, 0.2, 0.4) = 0.4$
*   $C_{13} = \max(\min(0.9, 0.4), \min(0.2, 0.5), \min(0.5, 1.0)) = \max(0.4, 0.2, 0.5) = 0.5$

**Row 2 of R1 against columns of R3:**
*   $C_{21} = \max(\min(0.4, 1.0), \min(0.8, 0.3), \min(0.1, 0.6)) = \max(0.4, 0.3, 0.1) = 0.4$
*   $C_{22} = \max(\min(0.4, 0.2), \min(0.8, 1.0), \min(0.1, 0.4)) = \max(0.2, 0.8, 0.1) = 0.8$
*   $C_{23} = \max(\min(0.4, 0.4), \min(0.8, 0.5), \min(0.1, 1.0)) = \max(0.4, 0.5, 0.1) = 0.5$

**Row 3 of R1 against columns of R3:**
*   $C_{31} = \max(\min(0.3, 1.0), \min(0.6, 0.3), \min(1.0, 0.6)) = \max(0.3, 0.3, 0.6) = 0.6$
*   $C_{32} = \max(\min(0.3, 0.2), \min(0.6, 1.0), \min(1.0, 0.4)) = \max(0.2, 0.6, 0.4) = 0.6$
*   $C_{33} = \max(\min(0.3, 0.4), \min(0.6, 0.5), \min(1.0, 1.0)) = \max(0.3, 0.5, 1.0) = 1.0$

**Final Matrix for R1 o R3 (Max-Min):**
$$ \begin{bmatrix} 0.9 & 0.4 & 0.5 \\ 0.4 & 0.8 & 0.5 \\ 0.6 & 0.6 & 1.0 \end{bmatrix} $$

---

## 4. Calculation: R1 o R3 (Max-Product Composition)

**Row 1 of R1 against columns of R3:**
*   $P_{11} = \max((0.9 \times 1.0), (0.2 \times 0.3), (0.5 \times 0.6)) = \max(0.90, 0.06, 0.30) = 0.90$
*   $P_{12} = \max((0.9 \times 0.2), (0.2 \times 1.0), (0.5 \times 0.4)) = \max(0.18, 0.20, 0.20) = 0.20$
*   $P_{13} = \max((0.9 \times 0.4), (0.2 \times 0.5), (0.5 \times 1.0)) = \max(0.36, 0.10, 0.50) = 0.50$

**Row 2 of R1 against columns of R3:**
*   $P_{21} = \max((0.4 \times 1.0), (0.8 \times 0.3), (0.1 \times 0.6)) = \max(0.40, 0.24, 0.06) = 0.40$
*   $P_{22} = \max((0.4 \times 0.2), (0.8 \times 1.0), (0.1 \times 0.4)) = \max(0.08, 0.80, 0.04) = 0.80$
*   $P_{23} = \max((0.4 \times 0.4), (0.8 \times 0.5), (0.1 \times 1.0)) = \max(0.16, 0.40, 0.10) = 0.40$

**Row 3 of R1 against columns of R3:**
*   $P_{31} = \max((0.3 \times 1.0), (0.6 \times 0.3), (1.0 \times 0.6)) = \max(0.30, 0.18, 0.60) = 0.60$
*   $P_{32} = \max((0.3 \times 0.2), (0.6 \times 1.0), (1.0 \times 0.4)) = \max(0.06, 0.60, 0.40) = 0.60$
*   $P_{33} = \max((0.3 \times 0.4), (0.6 \times 0.5), (1.0 \times 1.0)) = \max(0.12, 0.30, 1.00) = 1.00$

**Final Matrix for R1 o R3 (Max-Product):**
$$ \begin{bmatrix} 0.90 & 0.20 & 0.50 \\ 0.40 & 0.80 & 0.40 \\ 0.60 & 0.60 & 1.00 \end{bmatrix} $$

---

## 5. Calculation: R1 o (R2 o R3) (Max-Min Composition)

*Method: We first calculate R2 o R3 to create intermediate Matrix M, then calculate R1 o M.*

**Part A: Calculating M = R2 o R3 (Max-Min)**
*   **Row 1 of R2:**
    *   $M_{11} = \max(\min(0.7, 1.0), \min(0.3, 0.3), \min(0.6, 0.6)) = \max(0.7, 0.3, 0.6) = 0.7$
    *   $M_{12} = \max(\min(0.7, 0.2), \min(0.3, 1.0), \min(0.6, 0.4)) = \max(0.2, 0.3, 0.4) = 0.4$
    *   $M_{13} = \max(\min(0.7, 0.4), \min(0.3, 0.5), \min(0.6, 1.0)) = \max(0.4, 0.3, 0.6) = 0.6$
*   **Row 2 of R2:**
    *   $M_{21} = \max(\min(0.2, 1.0), \min(0.9, 0.3), \min(0.4, 0.6)) = \max(0.2, 0.3, 0.4) = 0.4$
    *   $M_{22} = \max(\min(0.2, 0.2), \min(0.9, 1.0), \min(0.4, 0.4)) = \max(0.2, 0.9, 0.4) = 0.9$
    *   $M_{23} = \max(\min(0.2, 0.4), \min(0.9, 0.5), \min(0.4, 1.0)) = \max(0.2, 0.5, 0.4) = 0.5$
*   **Row 3 of R2:**
    *   $M_{31} = \max(\min(0.5, 1.0), \min(0.1, 0.3), \min(0.8, 0.6)) = \max(0.5, 0.1, 0.6) = 0.6$
    *   $M_{32} = \max(\min(0.5, 0.2), \min(0.1, 1.0), \min(0.8, 0.4)) = \max(0.2, 0.1, 0.4) = 0.4$
    *   $M_{33} = \max(\min(0.5, 0.4), \min(0.1, 0.5), \min(0.8, 1.0)) = \max(0.4, 0.1, 0.8) = 0.8$

**Part B: Calculating R1 o M (Max-Min)**
*   **Row 1 of R1:**
    *   $C_{11} = \max(\min(0.9, 0.7), \min(0.2, 0.4), \min(0.5, 0.6)) = \max(0.7, 0.2, 0.5) = 0.7$
    *   $C_{12} = \max(\min(0.9, 0.4), \min(0.2, 0.9), \min(0.5, 0.4)) = \max(0.4, 0.2, 0.4) = 0.4$
    *   $C_{13} = \max(\min(0.9, 0.6), \min(0.2, 0.5), \min(0.5, 0.8)) = \max(0.6, 0.2, 0.5) = 0.6$
*   **Row 2 of R1:**
    *   $C_{21} = \max(\min(0.4, 0.7), \min(0.8, 0.4), \min(0.1, 0.6)) = \max(0.4, 0.4, 0.1) = 0.4$
    *   $C_{22} = \max(\min(0.4, 0.4), \min(0.8, 0.9), \min(0.1, 0.4)) = \max(0.4, 0.8, 0.1) = 0.8$
    *   $C_{23} = \max(\min(0.4, 0.6), \min(0.8, 0.5), \min(0.1, 0.8)) = \max(0.4, 0.5, 0.1) = 0.5$
*   **Row 3 of R1:**
    *   $C_{31} = \max(\min(0.3, 0.7), \min(0.6, 0.4), \min(1.0, 0.6)) = \max(0.3, 0.4, 0.6) = 0.6$
    *   $C_{32} = \max(\min(0.3, 0.4), \min(0.6, 0.9), \min(1.0, 0.4)) = \max(0.3, 0.6, 0.4) = 0.6$
    *   $C_{33} = \max(\min(0.3, 0.6), \min(0.6, 0.5), \min(1.0, 0.8)) = \max(0.3, 0.5, 0.8) = 0.8$

**Final Matrix for R1 o R2 o R3 (Max-Min):**
$$ \begin{bmatrix} 0.7 & 0.4 & 0.6 \\ 0.4 & 0.8 & 0.5 \\ 0.6 & 0.6 & 0.8 \end{bmatrix} $$

---

## 6. Calculation: R1 o (R2 o R3) (Max-Product Composition)

*Method: We first calculate R2 o R3 to create intermediate Matrix M, then calculate R1 o M.*

**Part A: Calculating M = R2 o R3 (Max-Product)**
*   **Row 1 of R2:**
    *   $M_{11} = \max((0.7 \times 1.0), (0.3 \times 0.3), (0.6 \times 0.6)) = \max(0.70, 0.09, 0.36) = 0.70$
    *   $M_{12} = \max((0.7 \times 0.2), (0.3 \times 1.0), (0.6 \times 0.4)) = \max(0.14, 0.30, 0.24) = 0.30$
    *   $M_{13} = \max((0.7 \times 0.4), (0.3 \times 0.5), (0.6 \times 1.0)) = \max(0.28, 0.15, 0.60) = 0.60$
*   **Row 2 of R2:**
    *   $M_{21} = \max((0.2 \times 1.0), (0.9 \times 0.3), (0.4 \times 0.6)) = \max(0.20, 0.27, 0.24) = 0.27$
    *   $M_{22} = \max((0.2 \times 0.2), (0.9 \times 1.0), (0.4 \times 0.4)) = \max(0.04, 0.90, 0.16) = 0.90$
    *   $M_{23} = \max((0.2 \times 0.4), (0.9 \times 0.5), (0.4 \times 1.0)) = \max(0.08, 0.45, 0.40) = 0.45$
*   **Row 3 of R2:**
    *   $M_{31} = \max((0.5 \times 1.0), (0.1 \times 0.3), (0.8 \times 0.6)) = \max(0.50, 0.03, 0.48) = 0.50$
    *   $M_{32} = \max((0.5 \times 0.2), (0.1 \times 1.0), (0.8 \times 0.4)) = \max(0.10, 0.10, 0.32) = 0.32$
    *   $M_{33} = \max((0.5 \times 0.4), (0.1 \times 0.5), (0.8 \times 1.0)) = \max(0.20, 0.05, 0.80) = 0.80$

**Part B: Calculating R1 o M (Max-Product)**
*   **Row 1 of R1:**
    *   $P_{11} = \max((0.9 \times 0.70), (0.2 \times 0.27), (0.5 \times 0.50)) = \max(0.63, 0.054, 0.25) = 0.63$
    *   $P_{12} = \max((0.9 \times 0.30), (0.2 \times 0.90), (0.5 \times 0.32)) = \max(0.27, 0.18, 0.16) = 0.27$
    *   $P_{13} = \max((0.9 \times 0.60), (0.2 \times 0.45), (0.5 \times 0.80)) = \max(0.54, 0.09, 0.40) = 0.54$
*   **Row 2 of R1:**
    *   $P_{21} = \max((0.4 \times 0.70), (0.8 \times 0.27), (0.1 \times 0.50)) = \max(0.28, 0.216, 0.05) = 0.28$
    *   $P_{22} = \max((0.4 \times 0.30), (0.8 \times 0.90), (0.1 \times 0.32)) = \max(0.12, 0.72, 0.032) = 0.72$
    *   $P_{23} = \max((0.4 \times 0.60), (0.8 \times 0.45), (0.1 \times 0.80)) = \max(0.24, 0.36, 0.08) = 0.36$
*   **Row 3 of R1:**
    *   $P_{31} = \max((0.3 \times 0.70), (0.6 \times 0.27), (1.0 \times 0.50)) = \max(0.21, 0.162, 0.50) = 0.50$
    *   $P_{32} = \max((0.3 \times 0.30), (0.6 \times 0.90), (1.0 \times 0.32)) = \max(0.09, 0.54, 0.32) = 0.54$
    *   $P_{33} = \max((0.3 \times 0.60), (0.6 \times 0.45), (1.0 \times 0.80)) = \max(0.18, 0.27, 0.80) = 0.80$

**Final Matrix for R1 o R2 o R3 (Max-Product):**
$$ \begin{bmatrix} 0.63 & 0.27 & 0.54 \\ 0.28 & 0.72 & 0.36 \\ 0.50 & 0.54 & 0.80 \end{bmatrix} $$